### PR TITLE
Cosine Warm Restarts: SGDR cyclical LR for multi-basin exploration

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The baseline uses `CosineAnnealingLR(T_max=160)` — a single cosine decay from peak LR to near-zero. The model converges into ONE basin and stays there. **Cosine Annealing with Warm Restarts (SGDR)** (Loshchilov & Hutter, ICLR 2017) periodically resets the LR to its peak value, giving the optimizer multiple chances to escape the current basin and explore different regions of the loss landscape.

The schedule: `T_0=40, T_mult=2` creates cycles of length 40, 80, 160 epochs:
- Epochs 0-40: first exploration cycle
- Epoch 40: LR resets to peak → escape current basin
- Epochs 40-120: second exploration cycle (80 epochs, deeper convergence)
- Epoch 120: LR resets → final exploration
- Epochs 120-280: third cycle (but training ends at ~150 epochs due to timeout)

The EMA model (decay=0.999) continuously averages across all cycles, effectively creating an implicit ensemble across multiple basins.

**Why it might work here:**
1. With only ~145-155 epochs before timeout, the single cosine decay barely reaches near-zero LR. Warm restarts give more time at high LR (exploration) in the early cycles.
2. The EMA model benefits from weight diversity across cycles — different basins contribute different strengths (some better for p_in, some for p_oodc).
3. OOD generalization often improves with flatter optima. Warm restarts prevent the optimizer from settling into a sharp minimum.

**Key distinction from SWA (#2234, fern's other experiment):**
- SWA: switches to constant LR + uniform weight averaging in the last 20%
- Warm restarts: cyclical LR throughout training + EMA averaging
- These are complementary approaches to the same goal (wider optima)

**Confidence:** Medium. SGDR is well-established in vision (DeiT, ViT). The risk is that the first cycle (40 epochs) is too short for this dataset, and restarts may undo useful learned representations.

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flags

```python
cosine_warm_restarts: bool = False   # SGDR: cosine annealing with warm restarts
cosine_T_0: int = 40                 # first cycle length in epochs
cosine_T_mult: int = 2              # cycle length multiplier (T_0, T_0*T_mult, T_0*T_mult^2, ...)
```

### Step 2: Replace scheduler

Find where `CosineAnnealingLR` is created. When `--cosine_warm_restarts` is enabled, use `CosineAnnealingWarmRestarts` instead:

```python
if cfg.cosine_warm_restarts:
    scheduler = torch.optim.lr_scheduler.CosineAnnealingWarmRestarts(
        optimizer, 
        T_0=cfg.cosine_T_0,      # 40 epochs first cycle
        T_mult=cfg.cosine_T_mult,  # 2x each subsequent cycle
        eta_min=1e-6              # minimum LR (same as baseline)
    )
else:
    scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
        optimizer, T_max=cfg.cosine_T_max, eta_min=1e-6
    )
```

**IMPORTANT:** `CosineAnnealingWarmRestarts` steps per EPOCH (same as the baseline scheduler). Make sure `scheduler.step()` is called once per epoch, not per batch.

### Step 3: No other changes needed

The rest of the training pipeline (EMA, PCGrad, loss computation, etc.) stays exactly the same. Only the LR schedule changes.

### Step 4: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent frieren --wandb_name "frieren/warm-restarts-s42" \
  --wandb_group "round19/cosine-warm-restarts" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --cosine_warm_restarts --cosine_T_0 40 --cosine_T_mult 2

# Seed 73 — identical but --seed 73 --wandb_name "frieren/warm-restarts-s73"
```

**NOTE:** Do NOT include `--cosine_T_max 160` — the warm restarts scheduler uses `T_0` and `T_mult` instead. Including both flags may cause a conflict — make sure the code handles this correctly (use warm restarts scheduler when `--cosine_warm_restarts` is set, regardless of `--cosine_T_max`).

### Step 5: Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, baseline comparison, W&B run IDs.

Also report:
- The LR curve shape — verify warm restarts are working (LR should spike at epochs ~40 and ~120)
- Which epoch achieved the best validation metrics (before or after a restart?)
- Whether the EMA model or the online model performs better

## Baseline

Current best (PR #2213, Wake Deficit Feature, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| **p_tan** | **28.341** | **< 28.34** |
| p_re   | **6.300**  | < 6.30  |

W&B runs: `hgml7i2r` (seed 42), `qic03vrg` (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent frieren --wandb_name "frieren/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Metrics (mae_surf_p)

| Metric | Baseline (PR #2213) | Seed 42 | Seed 73 | 2-seed Avg | Delta |
|--------|---------------------|---------|---------|------------|-------|
| **p_in** | **11.979** | 13.559 | 13.288 | 13.424 | **+12.1%** |
| **p_tan** | **28.341** | 29.992 | 31.247 | 30.620 | **+8.0%** |
| **p_oodc** | **7.643** | 8.839 | 9.309 | 9.074 | **+18.7%** |
| **p_re** | **6.300** | 7.209 | 7.286 | 7.247 | **+15.0%** |

### W&B Run IDs
- Seed 42: `9lercyhb` — [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/9lercyhb)
- Seed 73: `36bky4nk` — [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/36bky4nk)

### Command used
```bash
cd cfd_tandemfoil && python train.py \
  --agent frieren --wandb_name "frieren/warm-restarts-s42" \
  --wandb_group "round19/cosine-warm-restarts" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --cosine_warm_restarts --cosine_T_0 40 --cosine_T_mult 2
```
(Seed 73 identical with `--seed 73 --wandb_name "frieren/warm-restarts-s73"`)

Peak memory: ~60GB (GPU 0), ~76GB (GPU 1)
Total epochs: 149 (s42), ~148 (s73) — timeout at 180 min

### LR curve behavior
Warm restarts worked correctly:
- Cycle 1: epochs 0-40 (cosine decay from 2e-4 to 1e-5)
- **Reset at epoch 40**: LR jumped back to 2e-4
- Cycle 2: epochs 40-120 (cosine decay from 2e-4 to 1e-5)
- **Reset at epoch 120**: LR jumped back to 2e-4
- Cycle 3: epochs 120-149 (training ended mid-cycle at high LR ~1.8e-4)

Best epoch: **120** (s42) — right at the second restart point, before the third cycle undid progress.

### What happened

**Negative result.** All metrics regressed, worst being p_oodc (+18.7%) and p_re (+15.0%).

**Why it failed:**

1. **Insufficient convergence time per cycle.** The first cycle (40 epochs) is too short for this dataset/model complexity. The baseline's single cosine over ~150 epochs allows the model to slowly converge into a good basin. With T_0=40, the model barely starts learning before LR resets to peak, erasing partial progress.

2. **Third cycle cut short.** With training ending at epoch ~149, the third cycle (which would run epochs 120-280) only completed ~30 epochs at high LR. The model was still in exploration phase when training ended. Best metrics were at epoch 120 — the checkpoint saved just before the last LR reset destroyed it.

3. **EMA doesn't fully rescue the reset damage.** While EMA (decay=0.999) does smooth over the resets, a decay of 0.999 means the EMA half-life is ~693 steps. With ~331 steps/epoch, that's only ~2 epochs. After a full LR reset, the online model diverges quickly at peak LR, and EMA can't keep up — it starts mixing in the diverged weights within 2-3 epochs.

4. **Lion optimizer + warm restarts may be a bad combination.** Lion uses sign-based updates (not magnitude-dependent like Adam). The abrupt LR reset creates large, signed gradient steps that overwrite learned parameters more aggressively than with Adam-family optimizers.

### Suggested follow-ups

- **Longer first cycle (T_0=80, T_mult=2):** Would give 80, 160 epoch cycles. With ~150 total epochs, only one restart at epoch 80, giving 70 epochs of convergence in the second cycle. This is closer to a "cosine with one mid-training restart" which is less destructive.
- **Higher EMA decay during restarts (e.g. 0.9999):** Would better preserve the pre-restart model during the high-LR exploration phase. The current 0.999 is too fast.
- **SWA instead of restarts:** Stochastic Weight Averaging (already being tested in PR #2234 by fern) may achieve the "wide optima" goal more safely — it averages checkpoints rather than re-exploring from high LR.
- **Alternatively, accept the baseline cosine schedule is near-optimal** for this 3-hour training budget. The 160-epoch T_max is well-tuned for the ~150 actual epochs.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2238 because ADVISOR: Closing — all metrics regressed (p_in +12.1%, p_oodc +18.7%, p_tan +8.0%, p_re +15.0%). Your analysis is excellent. The core issue: with ~150 epoch budget, T_0=40 gives only 40 epochs for the first cycle, and the third cycle (epoch 120+) gets cut short at high LR. The model was in exploration phase when training ended. Lion's sign-based updates amplify the reset damage compared to Adam-family optimizers. The baseline cosine schedule (T_max=160) is well-matched to the ~150 actual epoch budget. Your suggestion about accepting this is correct — the single cosine is near-optimal for our training budget..
